### PR TITLE
Color cycle handling

### PR DESCRIPTION
**Pull request by mdboom**
Replaces #5674.

This uses the color cycle in more places.

It also introduces `Cn` syntax for colors to directly address the Nth element in a color cycle.  This was chosen over the `[n]` syntax in #5674 because it makes it more obvious that it's color-specific and can only be used in places where colors can be used.

### CODE REVIEW

File: examples/pylab_examples/color_demo.py [hunk 1]
```diff
@@ -11,6 +11,9 @@
     4) as a string representing a floating point number
        from 0 to 1, corresponding to shades of gray.
 
+    5) as a special color "Cn", where n is a number 0-9 specifying the
```
Comment (QuLogic) [inline examples/pylab_examples/color_demo.py]: The sentence above should be updated to say 5 ways.

File: examples/pylab_examples/color_demo.py [hunk 2]
```diff
@@ -20,8 +23,8 @@
 #subplot(111, facecolor='#ababab')
 t = np.arange(0.0, 2.0, 0.01)
 s = np.sin(2*np.pi*t)
-plt.plot(t, s, 'y')
-plt.xlabel('time (s)', color='r')
+plt.plot(t, s, 'C0')
+plt.xlabel('time (s)', color='C0')
```
Comment (QuLogic) [inline examples/pylab_examples/color_demo.py]: I'm pretty sure this demo is intentionally trying not to pick `C0`.

Comment (mdboom) [inline examples/pylab_examples/color_demo.py]: I'm deliberately updating it to the new recommended approach...

File: lib/matplotlib/axes/_base.py [hunk 3]
```diff
@@ -113,9 +116,14 @@ def _process_plot_format(fmt):
                 raise ValueError(
                     'Illegal format string "%s"; two color symbols' % fmt)
             color = c
+        elif c == 'C' and i < len(chars) - 1:
+            color_cycle_number = int(chars[i + 1])
```
Comment (QuLogic) [inline lib/matplotlib/axes/_base.py]: Limited to 10 I guess, but what if you wanted to use something like Vega20 (not that we have that as a cycle at the moment.)

File: lib/matplotlib/colors.py [hunk 4]
```diff
@@ -67,6 +73,19 @@
 
 def is_color_like(c):
     'Return *True* if *c* can be converted to *RGB*'
+
+    # Special-case the N-th color cycle syntax, because it's parsing
```
Comment (QuLogic) [inline lib/matplotlib/colors.py]: it's -> its

File: lib/matplotlib/axes/_base.py (hunk 3 above)
Comment (mdboom) [inline lib/matplotlib/axes/_base.py]: Yes, we should probably document that limitation.  The problem with opening it up to more than 10 is that it makes backward compatibility much harder.  Should `C11` be the 12th color or the 2nd color with a triangle up marker?  If we did allow more than one digit, we'd force people to have to write `1C1`.

Since more than 10 colors is sort of an edge case, and ill-advised even then, I think it's fine to have this limitation.

**Comment by QuLogic** [comment]
I assume the test image was not intended to have changed colours.

File: examples/pylab_examples/color_demo.py (hunk 2 above)
Comment (QuLogic) [inline examples/pylab_examples/color_demo.py]: Right, but I mean the original does _not_ pick blue.

**Comment by mdboom** [comment]
> I assume the test image was not intended to have changed colours.

Indeed it was intentional.  That test is testing a linestyle-only color cycle.  Since it has no colors, it defaults to black, which is a more sensible default than the old blue.

File: examples/pylab_examples/color_demo.py (hunk 1 above)
Comment (mdboom) [inline examples/pylab_examples/color_demo.py]: Thanks for noticing.  Done.

File: examples/pylab_examples/color_demo.py (hunk 2 above)
Comment (mdboom) [inline examples/pylab_examples/color_demo.py]: I see -- good point.  I'll update.

File: lib/matplotlib/colors.py (hunk 4 above)
Comment (mdboom) [inline lib/matplotlib/colors.py]: Fixed.

File: lib/matplotlib/axes/_axes.py [hunk 5]
```diff
@@ -3846,7 +3853,10 @@ def scatter(self, x, y, s=None, c=None, marker='o', cmap=None, norm=None,
             if facecolors is not None:
                 c = facecolors
             else:
-                c = 'b'  # The original default
+                if rcParams['_internal.classic_mode']:
+                    c = 'b'  # The original default
+                else:
+                    c = self._get_patches_for_fill.get_next_color()
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: I am +.75 on this.

Comment (mdboom) [inline lib/matplotlib/axes/_axes.py]: vs. `c = 'C0'`?

Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: Yeah, that is the only other plausibly defensible position.

My only reservation is that we have pushed back on this suggestion for so long and that in most cases if you are not mapping the color you _probably_ want to be using `plot(x, y, 'o')` instead.

That said, I have no real protest with this.

File: lib/matplotlib/rcsetup.py [hunk 6]
```diff
@@ -867,7 +867,7 @@ def validate_animation_writer_path(p):
     ## patch props
     'patch.linewidth':   [None, validate_float_or_None],  # line width in points
     'patch.edgecolor':   ['k', validate_color],  # black
-    'patch.facecolor':   ['#1f77b4', validate_color],  # blue (first color in color cycle)
+    'patch.facecolor':   ['C0', validate_color],  # blue (first color in color cycle)
```
Comment (jenshnielsen) [inline lib/matplotlib/rcsetup.py]: Delete blue in the comment here

**Comment by efiring** [comment]
This is related to #6328.

**Comment by tacaswell** [comment]
backported to v2.x as e52435406d0a7a8a8a44a485a753d9ed3a069adc
